### PR TITLE
4337793: Mark non-serializable fields of java.security.cert.Certificate and CertPath

### PR DESCRIPTION
**Pull request by seanjmullan**
Please review this 20+ year old bug (!), which marks the non-serializable fields of Certificate and CertPath with the transient modifier. These classes use an alternate serialization mechanism by overriding the writeReplace method. However, the fields of each class were never marked as transient and as a result were incorrectly documented in the Serialized Form section of the javadoc.

CSR: https://bugs.openjdk.java.net/browse/JDK-8277128

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-4337793](https://bugs.openjdk.java.net/browse/JDK-4337793): Mark non-serializable fields of java.security.cert.Certificate and CertPath

### Reviewers
 * [Valerie Peng](https://openjdk.java.net/census#valeriep) (@valeriepeng - **Reviewer**)
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6392/head:pull/6392` \
`$ git checkout pull/6392`

Update a local copy of the PR: \
`$ git checkout pull/6392` \
`$ git pull https://git.openjdk.java.net/jdk pull/6392/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6392`

View PR using the GUI difftool: \
`$ git pr show -t 6392`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6392.diff">https://git.openjdk.java.net/jdk/pull/6392.diff</a>

</details>

### CODE REVIEW

**Review by valeriepeng — APPROVED** [review]

**Review by RogerRiggs — CHANGES_REQUESTED** [review]
The serialized form doc for both classes should say something about what is serialized.
Since it is using writeReplace, it can indicate that CertificateRep/CertPathRep is used instead (and the arguments).
likely you'll need to use the \@serial javadoc tag and check the generated javadoc to verify.

**Comment by seanjmullan** [comment]
> The serialized form doc for both classes should say something about what is serialized. Since it is using writeReplace, it can indicate that CertificateRep/CertPathRep is used instead (and the arguments). likely you'll need to use the @serial javadoc tag and check the generated javadoc to verify.

The `writeReplace` methods do have `@serial` tags, and they do show up in the Serialized Form of the javadoc, ex: https://docs.oracle.com/en/java/javase/17/docs/api/serialized-form.html#java.security.cert.Certificate

Is your comment more that these methods could more clearly specify what is returned?

**Comment by seanjmullan** [comment]
> > > The serialized form doc for both classes should say something about what is serialized. Since it is using writeReplace, it can indicate that CertificateRep/CertPathRep is used instead (and the arguments). likely you'll need to use the @serial javadoc tag and check the generated javadoc to verify.
> > 
> > 
> > The `writeReplace` methods do have `@serial` tags, and they do show up in the Serialized Form of the javadoc, ex: https://docs.oracle.com/en/java/javase/17/docs/api/serialized-form.html#java.security.cert.Certificate
> > Is your comment more that these methods could more clearly specify what is returned?
> 
> The `@java.io.Serial` annotation doesn't add any description to the generated javadoc or serialized form doc.
> 
> The javadoc in the serialized for says only: "Replace the Certificate to be serialized." For my purposes it would enough to reword it to say it returns a CertificateRep holding the type and data or something similar. The `@serial` javadoc tag might be useful but not necessary if the regular javadoc that is in the serialized form doc is concrete.

Sounds good, see the latest commit for improvements to `writeReplace` and a few other wording improvements in related methods/fields.

**Review by RogerRiggs — APPROVED** [review]

**Review by valeriepeng — APPROVED** [review]
